File: go/src/13.罗马数字转整数.go
/*
 * @lc app=leetcode.cn id=13 lang=golang
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
func romanToInt(s string) int {
	m := map[byte]int{
		'I': 1,
		// 'IV': 4,
		'V': 5,
		// 'IX': 9,
		'X': 10,
		// 'XL': 40,
		'L': 50,
		// 'XC': 90,
		'C': 100,
		// 'CD': 400,
		'D': 500,
		// 'CM': 900,
		'M': 1000,
	}
	
	size, res := len(s), 0
	for i := 0; i < size; i++ {
		if i < size - 1 && m[s[i]] < m[s[i + 1]] {
			res -= m[s[i]]
		} else {
			res += m[s[i]]
		}
	}

	return res
}
// @lc code=end

File: rust/src/13.罗马数字转整数.rs
/*
 * @lc app=leetcode.cn id=13 lang=rust
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut m = HashMap::new();
        m.insert('I', 1);
        m.insert('V', 5);
        m.insert('X', 10);
        m.insert('L', 50);
        m.insert('C', 100);
        m.insert('D', 500);
        m.insert('M', 1000);
    
        let mut res = 0;
        let mut index = 0;
        let n = s.len();
        while index < n {
            let cur = m.get(&chars[index]).unwrap_or(&0);
            if index + 1 < n && cur < m.get(&chars[index+1]).unwrap_or(&0) {
                res += m.get(&chars[index+1]).unwrap() - cur;
                index += 1;
            } else {
                res += cur;
            }
            index += 1;
        }
    
        return res
    }
}
// @lc code=end

